### PR TITLE
Alert UI with metrics builder

### PR DESCRIPTION
**Pull request by mindhash**
This PR introduces :
- new create/edit form for alert definitions
- options to choose from query builder, promql 
- translations for create/edit alert form

Issue ref: [#347](https://github.com/SigNoz/engineering-pod/issues/347#issuecomment-1170790634)
Backend PR: [#1346](https://github.com/SigNoz/signoz/pull/1346)

### CODE REVIEW

**Comment by mindhash** [comment]
I am looking on build pipeline failures but not sure about playwright failure.

File: frontend/src/container/FormAlertRules/BasicInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+const { TextArea } = Input;
+const { Option } = Select;
+
+function BasicInfo({
+	alertDef,
+	setAlertDef,
+	queryCategory,
+}: {
+	alertDef: AlertDef;
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/BasicInfo.tsx:24]: Can we define interface here

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	allowCategoryChange,
+}: QuerySectionProps): JSX.Element {
+	// init namespace for translations
+	const { t } = useTranslation('rules');
+
+	const handleQueryCategoryChange = (s: string): void => {
+		if (!allowCategoryChange) {
+			notification.error({
+				message:
+					'Changes to query category are not allowed. Please create a new rule if the rule condition has changed.',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:53]: Can we put this message in inext?

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		setMetricQueries({
+			...allQueries,
+		});
+	};
+
+	const addMetricQuery = useCallback(async () => {
+		if (Object.keys(metricQueries).length > 4) {
+			notification.error({
+				message:
+					'Unable to create query. You can create at max 5 queries and 1 formulae.',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:157]: Same here also

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+				items: [],
+			},
+			groupBy: [],
+			disabled: false,
+			expression: queryLabel,
+		};
+		setMetricQueries({ ...queries });
+	}, [getNextQueryLabel, metricQueries, setMetricQueries]);
+
+	const addFormula = useCallback(async () => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:182]: Can we remove async here

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+					{t('button_query')}
+				</QueryButton>
+				<QueryButton onClick={addFormula} icon={<PlusOutlined />}>
+					{t('button_formula')}
+				</QueryButton>
+			</ButtonContainer>
+		);
+	};
+
+	const renderPromqlUI = (): JSX.Element => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:218]: Can we use useMemo here

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+		return (
+			<PromqlSection promQueries={promQueries} setPromQueries={setPromQueries} />
+		);
+	};
+
+	const renderMetricUI = (): JSX.Element => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:224]: Same here also can we wrap this one in useMemo

File: frontend/src/container/FormAlertRules/index.tsx [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+			toFormulaQueries(
+				initialValue?.condition?.compositeMetricQuery?.builderQueries,
+			),
+		);
+		setPromQueries({
+			...initialValue?.condition?.compositeMetricQuery?.promQueries,
+		});
+	}, [initialValue]);
+
+	const onCancelHandler = useCallback(async () => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:82]: There is no need to async here

File: frontend/src/container/FormAlertRules/index.tsx [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+		history.replace(ROUTES.LIST_ALL_ALERT);
+	}, []);
+
+	const isFormValid = useCallback((): boolean => {
+		let retval = true;
+
+		if (!alertDef.alert || alertDef.alert === '') {
+			notification.error({
+				message: 'Error',
+				description: 'alert name is required',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:92]: Can we put this message in i18 json

File: frontend/src/container/FormAlertRules/index.tsx [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+		}
+
+		if (
+			queryCategory === PROMQL &&
+			(!promQueries || Object.keys(promQueries).length === 0)
+		) {
+			notification.error({
+				message: 'Error',
+				description:
+					'promql expression is required when query format is set to PromQL',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:104]: Same here

File: frontend/src/container/FormAlertRules/index.tsx [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+			return false;
+		}
+
+		if (
+			(queryCategory === QUERY_BUILDER && !metricQueries) ||
+			Object.keys(metricQueries).length === 0
+		) {
+			notification.error({
+				message: 'Error',
+				description: 'at least one metric condition is required',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:115]: Here also

File: frontend/src/container/FormAlertRules/index.tsx [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+			});
+			return false;
+		}
+
+		Object.keys(metricQueries).forEach((key) => {
+			if (metricQueries[key].metricName === '') {
+				retval = false;
+				notification.error({
+					message: 'Error',
+					description: `metric name is missing in ${metricQueries[key].name}`,
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:125]: Here also

File: frontend/src/container/FormAlertRules/index.tsx [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+				});
+			}
+		});
+
+		Object.keys(formulaQueries).forEach((key) => {
+			if (formulaQueries[key].expression === '') {
+				retval = false;
+				notification.error({
+					message: 'Error',
+					description: `expression is missing in ${formulaQueries[key].name}`,
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:135]: Here also

File: frontend/src/container/FormAlertRules/index.tsx [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+				});
+			}
+		});
+
+		return retval;
+	}, [alertDef, queryCategory, metricQueries, formulaQueries, promQueries]);
+
+	const onSaveHandler = useCallback(async () => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:143]: Use try catch with async

File: frontend/src/container/FormAlertRules/index.tsx [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+						? 'Rule created successfully'
+						: 'Rule edited successfully',
+			});
+			setTimeout(() => {
+				history.replace(ROUTES.LIST_ALL_ALERT);
+			}, 2000);
+		} else {
+			notifications.error({
+				message: 'Error',
+				description: response.error || 'failed to create or edit rule',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:184]: Please use i18 next json here

File: frontend/src/container/FormAlertRules/index.tsx [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+		queryCategory,
+		ruleId,
+		notifications,
+		alertDef,
+		metricQueries,
+		formulaQueries,
+		promQueries,
+	]);
+
+	const renderBasicInfo = (): JSX.Element => (
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:199]: Can we wrap in useMemo

File: frontend/src/container/FormAlertRules/index.tsx [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+} from './utils';
+
+function FormAlertRules({
+	formInstance,
+	initialValue,
+	ruleId,
+}: FormAlertRuleProps): JSX.Element {
+	// init namespace for translations
+	const { t } = useTranslation('rules');
+	const [notifications, Element] = notification.useNotification();
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/index.tsx:34]: There is no need to use this hook we can use notification directly from antd

File: frontend/src/container/FormAlertRules/labels/QueryChip.tsx [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+
+import { QueryChipContainer, QueryChipItem } from './styles';
+import { ILabelRecord } from './types';
+
+export default function QueryChip({
+	queryData,
+	onRemove,
+}: {
+	queryData: ILabelRecord;
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/QueryChip.tsx:10]: Please make a interface here

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+import { QueryChipItem, SearchContainer } from './styles';
+import { ILabelRecord } from './types';
+import { createQuery, flattenLabels, prepareLabels } from './utils';
+
+function LabelSelect({
+	onSetLabels,
+	initialValues,
+}: {
+	onSetLabels: (q: Labels) => void;
+	initialValues: Labels | undefined;
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:26]: Here also

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+				}
+				setCurrentVal('');
+			},
+			onValidateQuery: (): void => {
+				console.log('onValidateQuery:', currentVal);
+				if (currentVal === '') {
+					return;
+				}
+
+				console.log('staging', staging);
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:60]: Please remove console here

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+
+				const generatedQuery = createQuery([...staging, currentVal]);
+
+				console.log('generatedQuery:', generatedQuery);
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:64]: Here also

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+				console.log('onSelectLabelValue:', currentVal);
+				if (currentVal !== '') {
+					setStaging((prevState) => [...prevState, currentVal]);
+				} else {
+					return;
+				}
+				setCurrentVal('');
+			},
+			onValidateQuery: (): void => {
+				console.log('onValidateQuery:', currentVal);
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:55]: Here also

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+
+	useEffect(() => {
+		if (initialValues) setQueries(flattenLabels(initialValues));
+	}, [initialValues]);
+
+	const [state, send] = useMachine(ResourceAttributesFilterMachine, {
+		actions: {
+			onSelectLabelKey: () => {},
+			onSelectLabelValue: () => {
+				console.log('onSelectLabelValue:', currentVal);
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:46]: Can you please remove this one

File: frontend/src/container/FormAlertRules/labels/index.tsx [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+
+	const handleClose = (key: string): void => {
+		dispatchChanges(queries.filter((queryData) => queryData.key !== key));
+	};
+
+	const handleClearAll = (): void => {
+		Modal.confirm({
+			title: 'Confirm',
+			icon: <ExclamationCircleOutlined />,
+			content: 'This action will remove all the labels. Do you want to proceed?',
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/index.tsx:101]: Can we put this one in i18 json

File: frontend/src/container/FormAlertRules/labels/styles.ts [hunk 24]
```diff
@@ -0,0 +1,32 @@
+import { grey } from '@ant-design/colors';
+import { Tag } from 'antd';
+import styled from 'styled-components';
+
+export const SearchContainer = styled.div<{
+	isDarkMode: boolean;
+	disabled: boolean;
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/styles.ts:7]: Please make interface here

File: frontend/src/container/FormAlertRules/labels/utils.ts [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+		}
+	});
+
+	return recs;
+};
+
+export const prepareLabels = (recs: ILabelRecord[]): Labels => {
+	const labels: Labels = {};
+
+	// eslint-disable-next-line no-plusplus
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/labels/utils.ts:39]: Can we remove this eslint rule

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/FormAlertRules/QuerySection.tsx (hunk 5 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/QuerySection.tsx]: typescript is throwing me off with useMemo here. I am kind of short of time, will handle this in future release if thats alright.

File: frontend/src/container/FormAlertRules/QuerySection.tsx (hunk 6 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/QuerySection.tsx:210]: will handle in future releases, as typescript is throwing me off with use of useMemo

File: frontend/src/container/FormAlertRules/index.tsx (hunk 8 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/index.tsx]: done

File: frontend/src/container/FormAlertRules/index.tsx (hunk 10 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/index.tsx]: done

File: frontend/src/container/FormAlertRules/index.tsx (hunk 15 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/index.tsx:274]: facing an issue while memoizing. Will handle in next release

File: frontend/src/container/FormAlertRules/labels/QueryChip.tsx (hunk 17 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/labels/QueryChip.tsx:10]: done

File: frontend/src/container/FormAlertRules/labels/utils.ts (hunk 25 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/labels/utils.ts:39]: done

**Comment by mindhash** [comment]
@palashgdev resolved most issues except the following: will need a little more time, will send a separate pr

- useMemo (typescript is throwing off when useMemo is used)
- annotations type In graph component - need to dig into annotation plugin to get the right type in

File: frontend/src/container/FormAlertRules/QuerySection.tsx (hunk 2 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/QuerySection.tsx]: done

File: frontend/src/container/FormAlertRules/labels/index.tsx (hunk 23 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/labels/index.tsx]: done

**Comment by palashgdev** [comment]
> @palashgdev resolved most issues except the following: will need a little more time, will send a separate pr
> 
> * useMemo (typescript is throwing off when useMemo is used)
> * annotations type In graph component - need to dig into annotation plugin to get the right type in

for useMemo typescript will not throw an error probably the root cause might be different

**Comment by palashgdev** [comment]
also @mindhash can we keep changes in this PR only?

File: frontend/src/api/alerts/save.ts [hunk 26]
```diff
@@ -0,0 +1,17 @@
+import { ErrorResponse, SuccessResponse } from 'types/api';
+import { PayloadProps, Props } from 'types/api/alerts/save';
+
+import create from './create';
+import put from './put';
```
Comment (palashgdev) [inline frontend/src/api/alerts/save.ts:5]: Can we remove this one and move this logic in the component itself

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+					query: '',
+					stats: '',
+					name: 'A',
+					legend: '',
+					disabled: false,
+				},
+			});
+		}
+
+		console.log('setting category:', parseInt(s, 10));
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:52]: Can we remove this console log

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 28]
```diff
@@ -0,0 +1,301 @@
...
+			allQueries[queryIndex] = current;
+		}
+
+		setMetricQueries({
+			...allQueries,
+		});
+	};
+
+	const addMetricQuery = useCallback(async () => {
+		if (Object.keys(metricQueries).length > 5) {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:143]: Please remove async here

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+				items: [],
+			},
+			groupBy: [],
+			disabled: false,
+			expression: queryLabel,
+		};
+		setMetricQueries({ ...queries });
+	}, [t, getNextQueryLabel, metricQueries, setMetricQueries]);
+
+	const addFormula = useCallback(async () => {
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:171]: Can you please remove async here

File: frontend/src/container/FormAlertRules/QuerySection.tsx [hunk 30]
```diff
@@ -0,0 +1,301 @@
...
+						defaultActiveKey={EQueryType.QUERY_BUILDER.toString()}
+						activeKey={queryCategory.toString()}
+						onChange={handleQueryCategoryChange}
+						// {tabBarExtraContent={
+						// 	<span style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
+						// 		<TextToolTip
+						// 			{...{
+						// 				text: `Preview Chart: This will stage and run the current query in the tab. Use this option to get a preview of query result.`,
+						// 			}}
+						// 		/>
```
Comment (palashgdev) [inline frontend/src/container/FormAlertRules/QuerySection.tsx:273]: Please remove this commented code

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/api/alerts/save.ts (hunk 26 above)
Comment (mindhash) [reply frontend/src/api/alerts/save.ts:5]: I think this makes the formAlertRules cleaner and separation of concern. It doesn't need to have multiple api calls. 

Is there any specific issue you see with this?

File: frontend/src/container/FormAlertRules/QuerySection.tsx (hunk 28 above)
Comment (mindhash) [reply frontend/src/container/FormAlertRules/QuerySection.tsx:142]: done

**Comment by mindhash** [comment]
@palashgdev resolved most except one. Please review and let me know if we can proceed.

**Review by palashgdev — DISMISSED** [review]
LGTM

**Pull request merged**